src/Npgsql/Internal/NpgsqlReadBuffer.cs

```diff
@@ -231,17 +231,24 @@ internal Task Ensure(int count, bool async, bool readingNotifications)
                             !connector.PostgresCancellationPerformed &&
                             connector.PerformPostgresCancellation())
                         {
-                            // Note that if the cancellation timeout is negative, we flow down and break the connection immediately
-                            var cancellationTimeout = connector.Settings.CancellationTimeout;
-                            if (cancellationTimeout >= 0)
+                            // SslStream treats any IOException (including timeouts) as fatal and may return garbage if reused.
+                            // To prevent this, we flow down and break the connection immediatelly.
+                            // See #4305.
+                            // The check below is equal to !(connector.IsSecure && e is IOException)
```

Comment:
Seems obvious enough that it doesn't need a comment :) Or we can extract out the negation if we prefer to...
```suggestion
```